### PR TITLE
Feature/create pair generator service

### PR DESCRIPTION
**Pull request by hromovyi-kostiantyn**
Added defaultProxy logic to DefaultPairGeneratorService

### CODE REVIEW

**Review by OlexTul — APPROVED** [review]

File: src/main/java/executor/service/service/parallelflowexecutor/impls/subscribers/DefaultPairGeneratorService.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+
+    @Override
+    public List<Pair> generatePairs() {
+        List<Pair> pairs = new ArrayList<>();
+        List<ProxyConfigHolder> proxies;
+        List<Scenario> scenarios;
+        try {
+            proxies = proxyQueue.getAllProxy();
+            scenarios = scenarioQueue.getAllScenario();
```
Comment (Sokol3783) [inline src/main/java/executor/service/service/parallelflowexecutor/impls/subscribers/DefaultPairGeneratorService.java]: Why do you declare variable in one place and initialize in another?

File: src/main/java/executor/service/service/parallelflowexecutor/impls/subscribers/DefaultPairGeneratorService.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            if (!scenarios.isEmpty() && !proxies.isEmpty()) {
+                if (proxies.size() == scenarios.size()) {
+                    pairs = createOneToOnePairs(proxies, scenarios);
+                } else {
+                    pairs = createMixPairs(proxies, scenarios);
+                }
+            } else if (!scenarios.isEmpty()) {
+                proxies.add(defaultProxy);
+                pairs = createMixPairs(proxies, scenarios);
```
Comment (Sokol3783) [inline src/main/java/executor/service/service/parallelflowexecutor/impls/subscribers/DefaultPairGeneratorService.java]: it's better create separate method, it's going be more clear and efficient

File: src/main/java/executor/service/service/parallelflowexecutor/impls/subscribers/DefaultPairGeneratorService.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        List<Scenario> scenarios;
+        try {
+            proxies = proxyQueue.getAllProxy();
+            scenarios = scenarioQueue.getAllScenario();
+            if (!scenarios.isEmpty() && !proxies.isEmpty()) {
+                if (proxies.size() == scenarios.size()) {
+                    pairs = createOneToOnePairs(proxies, scenarios);
+                } else {
+                    pairs = createMixPairs(proxies, scenarios);
+                }
```
Comment (Sokol3783) [inline src/main/java/executor/service/service/parallelflowexecutor/impls/subscribers/DefaultPairGeneratorService.java]: Can you hide this if/else in one function with common signature, and like this (but try to find out better name) :
return combineProxiesAndScenarios(proxies, scenario);

File: src/main/java/executor/service/service/parallelflowexecutor/impls/subscribers/DefaultPairGeneratorService.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            } else if (!scenarios.isEmpty()) {
+                proxies.add(defaultProxy);
+                pairs = createMixPairs(proxies, scenarios);
+            }
```
Comment (Sokol3783) [inline src/main/java/executor/service/service/parallelflowexecutor/impls/subscribers/DefaultPairGeneratorService.java]: There are four main cases:
1. scenarios and proxy has the same size
2. scenarios and proxy has different size, and both of them more than zero
3. there are no proxy
4. there are no scenarios

you describe three cases, but what will happen with proxy if there no scenarios?

**Review by Sokol3783 — CHANGES_REQUESTED** [review]

**Review by IfLunatic — APPROVED** [review]

**Review by Mercurialf — APPROVED** [review]

**Review by Sokol3783 — APPROVED** [review]

**Pull request merged**